### PR TITLE
[ERE-2119] Add "Unsubscribe all" email preference value to reporting

### PR DESCRIPTION
**Pull request by Lexiel46**

### CODE REVIEW

**Review by ScriptSmith — APPROVED** [review]
Looks good. Worth looking at reporting on individual `EmailNotificationPreference`s in a follow-up ticket as well?

**Review by ppettitau — CHANGES_REQUESTED** [review]
This looks great, well done!

In line with Adam's comment, I think if we want to introduce the other Email Preference details then we should 'future proof' the Unsubscribe All option by nesting it within an `emailPreferences` structure, similar to patient address and working groups, etc. If we released the Unsubscribe All in it's current position and then later refactored it, it would break any existing reports that had been created in the meantime.

@ScriptSmith , what do you think?
image

**Comment by ScriptSmith** [comment]
@ppettitau Good call, new option makes sense.

File: rdrf/report/report_configuration.py [hunk 1]
```diff
@@ -120,6 +118,19 @@ def get_patient_fields():
...
-            }
+            },
+
         }
     }
+
+    if EmailNotification.objects.is_registry_subscribable(all_registries):
+        demographic_model['demographic_model'].update({
+            'emailNotifications': {
+                'label': 'Email Notification',
```
Comment (ppettitau) [inline rdrf/report/report_configuration.py]: Rename to `'Email Notifications'` to keep consistent with naming of other reportable model labels

File: rdrf/rdrf/models/definition/models.py [hunk 2]
```diff
@@ -1157,6 +1157,9 @@ class EmailNotificationManager(models.Manager):
     def subscribable(self):
         return self.filter(subscribable=True)
 
+    def is_registry_subscribable(self, registries):
```
Comment (ppettitau) [inline rdrf/rdrf/models/definition/models.py]: Not necessary, but this function name feels a bit off to me.
Can you think of anything better e.g. `any_subscribable`, `has_any_subscribable` , `has_subscribable_registry`?
Also happy for it to be left as-is.

File: rdrf/report/schema.py [hunk 3]
```diff
@@ -489,6 +502,12 @@ def clinical_data_resolver(patient, _info):
             "resolve_clinical_data": clinical_data_resolver,
         })
```
Comment (ppettitau) [inline rdrf/report/schema.py:501]: Is this `if` condition necessary? I don't think anything will break if you leave it out

File: rdrf/report/report_configuration.py (hunk 1 above)
Comment (ppettitau) [reply rdrf/report/report_configuration.py]: Actually, it should be `'Email Preferences'`, when thinking about it from a Patient/ User perspective, we are getting data on their email preferences, not notifications.

File: rdrf/report/schema.py [hunk 4]
```diff
@@ -461,9 +470,13 @@ def clinical_data_resolver(patient, _info):
             collection="cdes"
         ).order_by('created_at').all()
 
+    def email_notification_resolver(patient, _info):
```
Comment (ppettitau) [inline rdrf/report/schema.py]: You can resolve the EmailPreference object to be consumed by the DjangoObjectType  from here
e.g. `return EmailPreference.objects.get_by_user(patient.user)`

File: rdrf/report/schema.py [hunk 5]
```diff
@@ -190,6 +191,14 @@ def resolve_gender(parent_guardian, _info):
         return dict(ParentGuardian.GENDER_CHOICES).get(parent_guardian.gender, parent_guardian.gender)
 
 
+class EmailNotificationsType(ObjectType):
+    unsubscribe_all = graphene.Boolean()
+
+    def resolve_unsubscribe_all(patient, _info):
```
Comment (ppettitau) [inline rdrf/report/schema.py]: As long as you change the resolver to resolve the EmailPreference, you can remove the lookup here and replace the `patient` parameter with the `emailPreference` object

**Review by ppettitau — APPROVED** [review]

**Review by ScriptSmith — APPROVED** [review]

**Pull request merged**